### DISCUSSION TITLE
VRS 2.x and Derivative Molecules

### DISCUSSION THREAD

**Discussion by ahwagner** [ga4gh/vrs#461] [Ideas]
The current Haplotype proposal for VRS 2.0 allows Adjacencies to accompany Alleles in the Haplotype `members`. Since Adjancency order is meaningful, the haplotype members array is now `ordered`. However, this requires some additional rules around expected order of Alleles.

@rrfreimuth recommended that we consider a model that does not impose ordering constraints on Alleles, perhaps as separate models. We are opening discussion on this point to collect opinions and ideas.

**Comment by d-cameron** [comment]
(partial) derivate chromosome (i.e. haplotype) reconstruction does require variant ordering for it to be unambiguous. We could support both ordered and unordered haplotypes but then we're changing our definition of a haplotype.

From my perspective, unordered is just an implicit linear ordering of the Alleles with respect to the reference. Is this not the case? When SVs are present, you can't assume that the standard intuitive textbook definition of one maternal + one paternal copy = 2 haplotypes. So what does haplotype actually mean then? For example, are de novo mutations on 2 different maternal copies of a trisomy chromosome considered the same haplotype? (No) What if there's one maternal chromosome but the region has been duplicated? (Yes?) Do the variants in the same duplicated region belong on the same haplotype or on different haplotypes? (same). None of this is made explicit in VRS and, given we have SV support, we need to define haplotype in more detail.

`Haplotypes are a specific combination of Alleles that are in-cis: occurring on the same physical molecule.`

Note that our existing definition of haplotype as based on the same physical molecule means that all VRS haplotypes implicitly require that the caller has verified the absence of any structural rearrangement that could separate the variants onto different physical molecules. For example, if you have fully phased (T2T) SNVs & a balanced chromosomal translocation is present, you need to break up your haplotype blocks because they're on different physical molecules. Is this what we intend? There are other definitions out there. If I say I've got fully phased maternal/paternal haplotypes then I'm making a claim about the inheritance pattern of those variants, not about the relationship between the physical molecules they occur on. Without SVs, these two definitions mean the same thing. With SVs, they do not.

**Reply by ahwagner** [reply]
Hey @d-cameron, I agree with most of what you're saying here and appreciate you taking the time to write this all out. I apologize for the delay in my response, but this was a lengthy post and I wanted to wait until I had a large block of time to respond to all of it.

> From my perspective, unordered is just an implicit linear ordering of the Alleles with respect to the reference. Is this not the case?

Yes, Allele order is not meaningful in the current model, because of this implication. This means fewer constraints on use of the data model, and reduced potential for implementation error.

> When SVs are present, you can't assume that the standard intuitive textbook definition of one maternal + one paternal copy = 2 haplotypes. So what does haplotype actually mean then? For example, are de novo mutations on 2 different maternal copies of a trisomy chromosome considered the same haplotype? (No) What if there's one maternal chromosome but the region has been duplicated? (Yes?) Do the variants in the same duplicated region belong on the same haplotype or on different haplotypes? (same).

Fully agreed–I know we see the problem and anticipated use of the Haplotype class with SVs the same way.

> Note that our existing definition of haplotype as based on the same physical molecule means that all VRS haplotypes implicitly require that the caller has verified the absence of any structural rearrangement that could separate the variants onto different physical molecules. For example, if you have fully phased (T2T) SNVs & a balanced chromosomal translocation is present, you need to break up your haplotype blocks because they're on different physical molecules. Is this what we intend?

VRS is used to describe the state of a sequence; if you perform an assay and find that SNPs A, B, and C are all co-located on the same molecule (what I think of as _in-cis_), that's a Haplotype. If A, B, and C are typically found _in-cis_ on a fully-phased reference assembly, that does not imply they are _in-cis_ when present in an assayed sample. IMO, calls made about that sample should use the Haplotype model only to convey a message about A, B, and C being _in-cis_ in that sample. Along that line of reasoning, if there is insufficient evidence to support them being _in-cis_, and especially if there is evidence to the contrary (e.g. a translocation) that is known to separate them, then they should simply be reported as Alleles (or alternate Haplotype blocks as informed by the assay). My opinion is moot, though, as VRS does not convey _how_ the message was made or the context it is used in, but only about _what_ the message says. In VRS, the message conveyed is the [computational definition of Haplotype](https://vrs.ga4gh.org/en/stable/terms_and_model.html#haplotype:~:text=A%20set%20of%20non%2Doverlapping%20Allele%20members%20that%20co%2Doccur%20on%20the%20same%20molecule.):

```A set of non-overlapping Allele members that co-occur on the same molecule.```

> There are other definitions out there.

The computational definition was derived from the biological definitions at [ISOGG](https://isogg.org/wiki/Haplotype), [SequenceOntology](http://www.sequenceontology.org/browser/current_release/term/SO:0001024), and [GENO](https://ontobee.org/ontology/GENO?iri=http://purl.obolibrary.org/obo/GENO_0000871). The recurring themes of how it is used in those definitions include _in-cis_ presence on chromosomes and genetic linkage, but I think you have rightly pointed out through example that the VRS information model only conveys information about the _in-cis_ part. It is also relaxed enough to consider derived chromosome "haplotypes" that conflict with the referenced source definitions. I wonder if the use of a biological term as a computational data structure is part of the issue here.

I'm going to make a separate thread with a model proposal that addresses this.

**Comment by Mrinal-Thomas-Epic** [comment]
I'm in favor of a single, ordered haplotype model, but am interested in hearing more about the perceived benefits of creating a separate, unordered haplotype model. My primary concerns are that it would create unnecessary complexity and more situations where there are multiple ways to represent the same haplotype.

Even if we defined an unordered haplotype model, we can't escape the need for an implicit ordering if we want to unordered haplotypes to have consistent identifiers, regardless of the order in which their alleles are specified. I think in VRS 1.3 this was done by [ordering](https://vrs.ga4gh.org/en/stable/impl-guide/computed_identifiers.html#digest-serialization) arrays of digests by unicode character sets values.

**Reply by ahwagner** [reply]
Hey Mrinal. I agree that we should not have two models with a shared class name, e.g. Haplotype (ordered) and Haplotype (unordered). I'm not 100% sure that is what @rrfreimuth meant, though I will leave it to him to comment further on that.

I do think he had a good point, though, in that ordering Alleles in a Haplotype will add unnecessary implementation complexity and increase the chance of errors. I think there is an obvious solution here, though, and will craft a separate thread about it.

**Comment by ahwagner** [comment]
![Phasing and Haplotypes](https://github.com/ga4gh/vrs/assets/811065/7d3e590f-3617-489d-8fd1-244d4cca3de5)

A draft model of 4 proposed changes between VRS `1.x` to `2.x` to support structural variation:

1. The `Haplotype` class gets renamed to `CisPhasedBlock` to differentiate between this concept and the notions of ancestry and linkage typically associated with `Haplotype` (leading to some of the concerns stated in [this comment](https://github.com/ga4gh/vrs/discussions/461#discussioncomment-8334579)).
2. A new `DerivativeSequence` class is developed that allows construction of novel contiguous sequences from multiple adjacencies and CisPhasedBlocks.
3. The previously-described `Adjacency` class is added that allows for the representation of two adjacent sequences and an optional intervening linker sequence expression.
4. A new `SequenceTerminus` class is added that allows for the specification of a single sequence location, representing the end of a sequence.

These changes address the concerns about ordered / unordered variant members by creating data classes specifically for each of _cis_-phased variant and derivative sequences.

**Reply by ahwagner** [reply]
@larrybabb `DerivativeSequence` object should support `Allele` as well (for when there is _only_ one `Allele` and not a cis-phased block).

**Reply by cmprocknow** [reply]
Would this mean a CisPhasedBlock should have at least 2 Allele members?

**Reply by Mrinal-Thomas-Epic** [reply]
What would be the advantage of using CPB with multiple alleles in `DerivativeSequence`, rather than putting multiple `Alleles` in a row in `DerivativeSequence.components`?

What is the "Block" in `CisPhasedBlock`? Since the class only contains Alleles, could we call it `CisPhasedAlleles`?

**Reply by ahwagner** [reply]
@cmprocknow yes, a CisPhasedBlock should have at least 2 Allele members.

@Mrinal-Thomas-Epic:
>What would be the advantage of using CPB with multiple alleles in DerivativeSequence, rather than putting multiple Alleles in a row in DerivativeSequence.components?

There are three advantages to this approach. The first of these is semantics; the members of a Haplotype (VRS 1.3) / CPB (Draft VRS 2.0) represent allele co-occurrence on a molecule, and the components of a DerivativeSequence represent a derivate molecule composed from molecular adjacencies, which may include variants that reside on those component molecules. Separating out these classes allows us to semantically represent these distinct scenarios explicitly, instead of implicitly by the components used. As I write this, I wonder if we should be calling this class a `DerivativeMolecule` instead. 🤔

The second advantage is that by separating these classes out, implementations have no need to validate that order of elements matches a convention (e.g. ordering Alleles in increasing or decreasing coordinate space to agree with coordinates used by Adjacency components). Instead, order is always meaningful (or not meaningful) as defined by the containing object. This addresses the concern raised by @rrfreimuth.

The third advantage is that multiple Alleles in a CPB can be referenced as a unit in a way that is identical to the Haplotype model from VRS 1.3; meaning that the identifier for those CPBs can be used as a drop-in replacement in describing CPBs that exist on DerivedSequences.

I also put together an illustration to illustrate
![Phasing and Haplotypes - Rationale](https://github.com/ga4gh/vrs/assets/811065/a882fe94-da2b-48f5-b51b-ace544c450ab)
 these second and third points, hopefully this helps:


>What is the "Block" in CisPhasedBlock? Since the class only contains Alleles, could we call it CisPhasedAlleles?

I chose "Block" since an instance of the class should represent a singular entity (e.g. "A haplotype" / "A cis phased block", not "A cis phased alleles"). Block is used elsewhere in the community to refer to similar concepts, e.g. haplotype blocks [[1]](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6707469/#:~:text=Existing%20methods%20commonly%20define%20a%20haplotype%20block%20as%20a%20set%20of%20adjacent%20loci%2C%20using%20either%20a%20fixed%20number%20of%20markers%20or%20a%20fixed%20number%20of%20different%20sequences%20of%20alleles%20per%20block%20(Meuwissen%20et%20al.%202014)), [[2]](https://www.pacb.com/blog/ploidy-haplotypes-and-phasing/#:~:text=to%20phase%20large-,haplotype%20blocks,-.) and phase blocks [[3]](https://support.10xgenomics.com/genome-exome/software/visualization/latest/phase-blocks). Other possible terms for this unit could include "Set" (e.g. [VCF](https://samtools.github.io/hts-specs/VCFv4.4.pdf) Phase Set) or "Group" (generic).

If we think it is useful to forward-constrain this class to only Alleles, we can consider adding Alleles to the class name; this might look like `CisPhasedAlleleBlock` or something like that. The tradeoff for this approach is the extra message space it would use in the `type` field when representing these concepts as JSON, so I think we should make sure doing so would help disambiguate it from other envisioned VRS classes. My initial feeling is we should not change the name to include "Allele" as I do not see much perceived benefit to accompany the tradeoff, but if you (or anyone else reading this) has reason to think we should add Allele to the class name (or any other changes to the class name), please do indicate so here.

**Comment by larrybabb** [comment]
If you are looking at the 2.x vrs.ga4gh.org readthedocs specification the `DerivativeMolecule` Concept under `MolecularVariation` will point you back to this discussion.

**Comment by larrybabb** [comment]
@ahwagner As I'm converting the 1.3 sv_haplotype test I noticed that we do not have `Allele` as `oneOf` the options for the `DerivativeSequence.components` array. Based on my reading/understanding of the above, we must allow `Allele` to be a component when ordering of more than one `Allele` in a row is important or when there is only one `Allele` between `Adjacency`'s.  We require at least 2 members in a `CisPhasBlock` so a single `Allele` can NOT be described by a `CisPhaseBlock`.   Please verify if this is correct.

**Reply by ahwagner** [reply]
Consensus from 7/30 call was to also allow `Allele`.
